src/controller/CurrentFabricRemover.cpp

```diff
@@ -0,0 +1,190 @@
+
+    return mCommissioner->GetConnectedDevice(remoteDeviceId, &mOnDeviceConnectedCallback, &mOnDeviceConnectionFailureCallback);
+}
+
+CHIP_ERROR CurrentFabricRemover::ReadCurrentFabricIndex(Messaging::ExchangeManager & exchangeMgr, SessionHandle & sessionHandle)
+{
+    using TypeInfo = chip::app::Clusters::OperationalCredentials::Attributes::CurrentFabricIndex::TypeInfo;
+    OperationalCredentialsCluster cluster(exchangeMgr, sessionHandle, kRootEndpointId);
+
+    return cluster.ReadAttribute<TypeInfo>(this, OnSuccessReadCurrentFabricIndex, OnReadAttributeFailure);
+}
+
+CHIP_ERROR CurrentFabricRemover::SendRemoveFabricIndex(Messaging::ExchangeManager & exchangeMgr, SessionHandle & sessionHandle)
+{
+    if (mFabricIndex == kUndefinedFabricIndex)
+    {
+        return CHIP_ERROR_INVALID_FABRIC_INDEX;
+    }
+    ChipLogProgress(Controller, "SendRemoveFabricIndex : %u", mFabricIndex);
+
+    OperationalCredentials::Commands::RemoveFabric::Type request;
+    request.fabricIndex = mFabricIndex;
+
+    OperationalCredentialsCluster cluster(exchangeMgr, sessionHandle, 0);
+
+    return cluster.InvokeCommand(request, this, OnSuccessRemoveFabric, OnCommandFailure);
+}
+
+void CurrentFabricRemover::OnDeviceConnectedFn(void * context, Messaging::ExchangeManager & exchangeMgr,
+                                               SessionHandle & sessionHandle)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    auto * self    = static_cast<CurrentFabricRemover *>(context);
+    VerifyOrReturn(self != nullptr, ChipLogProgress(Controller, "Device connected callback with null context. Ignoring"));
+
+    switch (self->mNextStep)
+    {
+    case Step::kReadCurrentFabricIndex: {
+        err = self->ReadCurrentFabricIndex(exchangeMgr, sessionHandle);
+        break;
+    }
+    case Step::kSendRemoveFabric: {
+        err = self->SendRemoveFabricIndex(exchangeMgr, sessionHandle);
+        break;
+    }
+    default:
+        err = CHIP_ERROR_INCORRECT_STATE;
+        break;
+    }
+
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Controller, "%" CHIP_ERROR_FORMAT, err.Format());
+        FinishRemoveCurrentFabric(context, err);
+    }
+
+    return;
+}
+
+void CurrentFabricRemover::OnDeviceConnectionFailureFn(void * context, const ScopedNodeId & peerId, CHIP_ERROR err)
+{
+    ChipLogProgress(Controller, "OnDeviceConnectionFailureFn: %s", err.AsString());
+
+    auto * self = static_cast<CurrentFabricRemover *>(context);
+    VerifyOrReturn(self != nullptr, ChipLogProgress(Controller, "Device connected callback with null context. Ignoring"));
+
+    FinishRemoveCurrentFabric(context, err);
+    return;
+}
+
+void CurrentFabricRemover::OnSuccessReadCurrentFabricIndex(void * context, uint8_t fabricIndex)
+{
+    auto * self = static_cast<CurrentFabricRemover *>(context);
+    VerifyOrReturn(self != nullptr, ChipLogProgress(Controller, "Device connected callback with null context. Ignoring"));
```

Comment:
This is not a device connected callback.  The same copy/paste error occurs in a number of functions below.